notebooks/02_model/vowpal_wabbit_deep_dive.ipynb

```diff
@@ -0,0 +1,737 @@
+    "        output (str): path to output file\n",
+    "    \"\"\"\n",
+    "    with open(output, 'w') as f:\n",
+    "        tmp = df.reset_index()\n",
+    "\n",
+    "        # we need to reset the rating type to an integer to simplify the vw formatting\n",
+    "        tmp['rating'] = tmp['rating'].astype('int64')\n",
+    "        \n",
+    "        # convert rating to binary value\n",
+    "        if logistic:\n",
+    "            tmp['rating'] = tmp['rating'].apply(lambda x: 1 if x >= 3 else -1)\n",
+    "        \n",
+    "        for _, row in tmp.iterrows():\n",
+    "            f.write('{rating} {index}|u {userID} |i {itemID}\\n'.format_map(row))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def run_vw(train_params, test_params, test_data, prediction_path):\n",
+    "    \"\"\"Convenience function to train, test, and show metrics of interest\n",
+    "    Args:\n",
+    "        train_params (str): vw training parameters\n",
+    "        test_params (str): vw testing parameters\n",
+    "        test_data (pd.dataFrame): test data\n",
+    "        prediction_path (str): path to vw prediction output   \n",
+    "    \"\"\"\n",
+    "\n",
+    "    # train model\n",
+    "    run(train_params.split(' '), check=True)\n",
+    "    \n",
+    "    # test model\n",
+    "    run(test_params.split(' '), check=True)\n",
+    "    \n",
+    "    # read in predictions\n",
+    "    pred_data = pd.read_csv(prediction_path, delim_whitespace=True, names=['prediction'], index_col=1).join(test_data)\n",
+    "    \n",
+    "    # ensure results are integers in correct range\n",
+    "    pred_data['prediction'] = pred_data['prediction'].apply(lambda x: int(max(1, min(5, round(x)))))\n",
+    "\n",
+    "    for f in [rmse, mae, rsquared, exp_var]:\n",
+    "        print('{name}: {metric}'.format(name=f.__name__.upper(), metric=f(test_data, pred_data)))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# create temp directory to maintain data files\n",
+    "tmpdir = TemporaryDirectory()\n",
+    "\n",
+    "model_path = os.path.join(tmpdir.name, 'vw.model')\n",
+    "train_path = os.path.join(tmpdir.name, 'train.dat')\n",
+    "test_path = os.path.join(tmpdir.name, 'test.dat')\n",
+    "train_logistic_path = os.path.join(tmpdir.name, 'train_logistic.dat')\n",
+    "test_logistic_path = os.path.join(tmpdir.name, 'test_logistic.dat')\n",
+    "prediction_path = os.path.join(tmpdir.name, 'prediction.dat')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<h3>Load & Transform Data</h3>"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# load movielens data (use the 1M dataset)\n",
+    "df = load_pandas_df('1m')\n",
+    "\n",
```

Comment:
Just FYI @gramhagen I'm working on explaining how to do tests on notebooks with papermil: https://github.com/Microsoft/Recommenders/issues/455

Reply:
👍